Qpf/Macro/Data/Constructors.lean

```diff
@@ -0,0 +1,64 @@
+import Qpf.Macro.Data.Replace
+import Qpf.Macro.Data.View
+
+open Lean Meta Elab.Command
+open PrettyPrinter (delab)
+
+namespace Data.Command
+open Parser in
+/--
+  Count the number of arguments to a constructor
+-/
+partial def countConstructorArgs : Syntax → Nat
+  | Syntax.node _ ``Term.arrow #[_, _, tail]  =>  1 + (countConstructorArgs tail)
+  | _                                         => 0
+
+
+open Elab
+/--
+  Add convenient constructor functions to the environment
+-/
+def mkConstructors (view : DataView) (shape : Name) : CommandElabM Unit := do
+  for ctor in view.ctors do
+    trace[QPF] "mkConstructors\n{ctor.declName} : {ctor.type?}"
+    let n_args := (ctor.type?.map countConstructorArgs).getD 0
+
+    let args ← (List.range n_args).mapM fun _ =>
+      do pure <| mkIdent <|← Elab.Term.mkFreshBinderName
+    let args := args.toArray
```

Comment:
Stylistically, I don't like variable shadowing, as it seems to imply that the same variable name has two different purposes.